CanonicalABI.md

```diff
@@ -0,0 +1,1137 @@
+          yield(variant_start<$cases>(discrim))
+
+          # Recursively read the payload, if present for this case. Note that
+          # we align the pointer to the whole variant's alignment to ensure that
+          # all payloads start at the same address.
+          if payload:
+            ptr = align_to(ptr, align(direction, variant $cases))
+            yield(lift_from_memory(direction, dst, ptr, payload))
+
+          yield(variant_end<$cases>)
+        }
+      }
+    }
+
+    # defer validation and such to `lift`
+    handle<$resource> => {
+      val = read_4_bytes(src, ptr)
+      lift(direction, src, handle<$resource>, [val as wasm_i32])
+    }
+
+    push-buffer<$ty> | pull-buffer<$ty>(val) => {
+      assert(direction == "import") # see comments above
+      handle = read_4_bytes(src, ptr)
+      offset = read_4_bytes(src, ptr + 4)
+      len = read_4_bytes(src, ptr + 8)
+      lift(direction, src, ty, [handle as wasm_i32, offset as wasm_i32, len as wasm_i32])
+    }
+```
+
+## Lowering
+
+The `lower` function here takes three parameters:
+
+* `direction` - one of `"export"` or `"import"` which indicates for which
+  purposes the lowering is happening: lowering an export means lowering
+  a parameter, lowering an import means lowering a result.
+* `dst` - this is an instance where the value is being lowered into. This could
+  be a WebAssembly instance or the host itself.
+* `gen` - a generator representing the interface types values that's being
+  lowered. This is expected to be the generator returned from calling `lift`
+  above.
+
+This defines how an interface types value is encoded into linear memory and
+WebAssembly values.
+
+```
+def lower(direction, dst, gen):
+  match gen.next()
+    f32(val) => [val as wasm_f32]
+    f64(val) => [val as wasm_f64]
+
+    # sign-extended
+    s8(val) => [val as wasm_i32]
+    s16(val) => [val as wasm_i32]
+
+    # zero-extended
+    u8(val) => [val as wasm_i32]
+    u16(val) => [val as wasm_i32]
+
+    s32(val) => [val as wasm_i32]
+    u32(val) => [val as wasm_i32]
+    s64(val) => [val as wasm_i64]
+    u64(val) => [val as wasm_i64]
+
+    char(val) => [val as wasm_i32]
+
+    # Note that `val` here is guaranteed to be a valid list of unicode scalar
+    # values so encoding here should always succeed.
+    string(val) => {
+      base = dst.canonical_abi_malloc(utf8_byte_length(val), 1)
+      encode_utf8_string(dst, base, val)
+      [base as wasm_i32, val.utf8_len() as wasm_i32]
+    }
+
+    # List are represented by their pointer/length as 32-bit integers. Note that
+    # the `canonical_abi_malloc` can be elided if `dst` is the host itself since
+    # the host can retain the raw pointers into the wasm module's memory if it
+    # likes.
+    list_start<$ty>(len) => {
+      base = dst.canonical_abi_malloc(len * size(direction, $ty), align(direction, $ty))
+      assert_or_trap(base % align(direction, $ty) == 0)
```

Comment:
Similar for lifting, it'd probably be good to add an `assert_or_trap` for in-bounds before doing any effectful copying.